state/protocol/badger/state.go

```diff
@@ -112,9 +250,47 @@ func (s *State) AtBlockID(blockID flow.Identifier) protocol.Snapshot {
 	return NewSnapshot(s, blockID)
 }
 
-func (s *State) Mutate() protocol.Mutator {
-	m := &Mutator{
-		state: s,
+// newState initializes a new state backed by the provided a badger database,
+// mempools and service components.
+// The parameter `expectedBootstrappedState` indicates whether or not the database
+// is expected to contain a an already bootstrapped state or not
+func newState(
+	metrics module.ComplianceMetrics,
+	db *badger.DB,
+	headers storage.Headers,
+	seals storage.Seals,
+	blocks storage.Blocks,
+	setups storage.EpochSetups,
+	commits storage.EpochCommits,
+	statuses storage.EpochStatuses,
+) *State {
+	return &State{
+		metrics: metrics,
+		db:      db,
+		headers: headers,
+		seals:   seals,
+		blocks:  blocks,
+		epoch: struct {
+			setups   storage.EpochSetups
+			commits  storage.EpochCommits
+			statuses storage.EpochStatuses
+		}{
+			setups:   setups,
+			commits:  commits,
+			statuses: statuses,
+		},
+	}
+}
+
+// IsBootstrapped returns whether or not the database contains a bootstrapped state
+func IsBootstrapped(db *badger.DB) (bool, error) {
+	var finalized uint64
+	err := db.View(operation.RetrieveFinalizedHeight(&finalized))
```

Comment:
I'm curious what's your thoughts of changing from using [the protocol abstraction](https://github.com/onflow/flow-go/pull/236/files#diff-cae89c2f7daa439631988089cdd51c551e4c25db2452885527785d0774235041L427) (`Finalize().Head()`) to a concrete database method call?

Reply:
This was implemented by Alex, I will leave my thoughts, @AlexHentschel please comment if I am wrong. Previously we had next flow: 
1. call `badger.NewState` to construct object of `badger.State` in invalid state.
2. call `Bootstrap` or load state from DB.

At current implementation you can't create `badger.State` in some invalid state, you will construct it with `badger.OpenState` which loads everything from DB and creates a valid `badger.State` and `badger.StateRoot` or you can use `badger.Bootstrap` in case you don't have a DB. So this is a cleaner approach to constructing objects in atomic way. But for this you need a way to lookup DB to check if it's created or not, you can't use `Finalize().Head()` as previously since you don't have the `State` object.

Reply:
right, makes sense. the `State` object isn't available to be called yet 👍